### PR TITLE
SOLR-16962: Restore ability to configure tlog directory

### PR DESCRIPTION
**Pull request by magibney**
See: SOLR-16962

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/core/SolrCore.java
Comment (magibney) [inline solr/core/src/java/org/apache/solr/core/SolrCore.java:397]: My best understanding is that this was a kind of memoization, to prevent repeated computation of the same string? But it's misleading to have it in here at all, since the only context in which it was used did not actually need the _default_ value (this is available from UpdateLog instance), but rather was interested in consulting the explicit value from CoreDescriptor (often `null`), to properly determine ulog definition hierarchy:
1. from `core.properties`
2. from `updateLog` "dir" config
3. default (`[instanceDir]/data/tlog/`)

... so incorporating the third of the above hierarchy into the `SolrCore.getUlogDir()` abstraction made it useless in the only contexts in which it was actually used.

**Review by dsmiley — APPROVED** [review]
+1 (didn't review too closely this time).

One thing to watch out for with close hooks is to not accidentally re-register a close hook to do the same thing.  Not sure if that's an issue here but wanted to mention it.

**Comment by magibney** [comment]
This turned into a bigger change than I'd anticipated, but I really didn't see any other way around it. I found that I had to refactor (for clarity) some method/variable names to be more explicit about the distinction between tlog and ulog. Both concepts are indeed distinct, and are required for different purposes in different places.

`ulog` dir remains the directory that is configured by the `updateLog.dir` param, and is the _parent_ directory of the `tlog` dir. The `tlog` dir is the directory that contains the transaction log files. Specifying `ulog` dir as the parent directory allows consistent behavior for the normal case (where we want the `tlog` dir to be implicitly scoped to the core as a consequence of being a subdirectory within the core `data` dir (normally) or core instance dir), and the special case (where an external shared top-level `ulog` dir is configured, within which directories must be created that are explicitly scoped to their associated core).

I opted to track the tlog dirs through the core `DirectoryFactory`, which makes directory deletion safer (can't delete untracked directories) and more flexible (e.g., supporting (or at least not breaking for) hdfs).

**Review by justinrsweeney — APPROVED** [review]
This looks good to me and certainly seems like an improvement on what we had before.

**Comment by jkrauss82** [comment]
When is this going to be merged?

I am in progress of configuring a Solr Cloud cluster and was baffled the ulog dir is silently ignored by the Solr server.

I believe any serious production setup will keep data and tx logs on separate drives so I think this should be merged ASAP, given the documentation [clearly says](https://solr.apache.org/guide/solr/latest/configuration-guide/commits-transaction-logs.html#transaction-log-configuration) the ulog dir variable is considered.

**Comment by magibney** [comment]
Agreed, and thanks for the nudge @jkrauss82. This got too close to the 9.6 release and I didn't want to rush it in, but I'll go ahead and merge this today or tomorrow. I'm personally hoping for a 9.7 soon (esp. thinking also of https://github.com/apache/solr/pull/1557).

File: solr/core/src/test/org/apache/solr/update/CustomTLogDirTest.java
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/update/CustomTLogDirTest.java:43]: STCJ4 arranges for props to be cleared

**Review by dsmiley — COMMENTED** [review]
I hate to say it but before this is merged, the Hdfs tests ought to be in a passing state.  Otherwise we won't know if we broke something.  I looked at http://fucit.org/solr-jenkins-reports/failure-report.html and I suppose the situation isn't so terrible after all.  Any way, be sure to pay attention here after a merge.

If you're lucky, you'll get a code review from @markrmiller who had his hands all over this once upon a time.  Any way... hope & pray nothing breaks :-).  This part of Solr is a bit of a mystery.  I'm glad you are improving it.  CHANGES.txt might mention touching HDFS even though no expected functional changes.  It's a risk someone running HDFS might want to be aware of.

**Comment by magibney** [comment]
Thanks, yes I will definitely keep an eye on the builds.

I'll wait a few days -- @markrmiller if you're able to weigh in that would be much appreciated!

I do have a specific question about [this](https://github.com/apache/solr/blob/11141c0c6a788213e15a9eff1b5fe7e2ea4870ad/solr/core/src/java/org/apache/solr/update/DirectUpdateHandler2.java#L196):
```java
    UpdateLog existingLog = updateHandler.getUpdateLog();
    if (this.ulog != null && this.ulog == existingLog) {
      // If we are reusing the existing update log, inform the log that its update handler has
      // changed. We do this as late as possible.
      this.ulog.init(this, core);
    }
```
... specifically, why do we need to do this "as late as possible"? afaict the only thing the UpdateHandler is used for is to update the UpdateLog's ref to the UpdateHandler -- no fields, etc. consulted. This would be far simpler if we _always_ called `ulog.init(this, core)` from the super ctor (`UpdateHandler`), as long as `ulog != null`.

I'm also particularly interested in feedback on [this TODO](https://github.com/apache/solr/blob/962b99f430c56f7387fbf233950fbc12946aaa70/solr/core/src/java/org/apache/solr/update/UpdateLog.java#L431-L436):
```java
      // TODO: why do we need fancy concurrency constructs here? I don't see where this would
      //  actually be called concurrently. Could `TestHdfsUpdateLog.testFSThreadSafety()`
      //  (introduced by SOLR-7113) in fact be the only place that requires this? Or perhaps we
      //  need the `init(UpdateHandler, SolrCore)` method in its entirety (or some large portion of
      //  it) to synchronize on `this`, to ensure visibility of non-final fields that are
      //  initialized?
```

**Comment by magibney** [comment]
The build failure for 962b99f430c56f7387fbf233950fbc12946aaa70 is legit, but not related to this PR. I've opened an [issue](https://issues.apache.org/jira/browse/SOLR-17323)/[PR](https://github.com/apache/solr/pull/2505) for that failure.

**Comment by magibney** [comment]
For the record, considering the above question about concurrency constructs:

I've dug into [SOLR-7113](https://issues.apache.org/jira/browse/SOLR-7113), and I'm increasingly convinced that although it's technically correct that `UpdateLog.init(UpdateHandler, SolrCore)` is not thread safe, the issue title (filed with initial report) is misleading in calling attention to "thread safety", and the issue fix (in the form of making a portion of the method synchronized on `fsLock` object) is misleading, because afaict `UpdateLog.init(UpdateHandler, SolrCore)` is never called concurrently in application code, nor should it be.

The issue with SOLR-7113 was not one of thread safety, so much as one of leaking resources and closing them across different scopes. This issue was effectively fixed by disallowing modification of tlog location across core reloads.

My inclination atm is to leave the AtomicBoolean in place -- we need something to shortcircuit repeated init in application code (which does happen, albeit not in a way that needs to be thread safe), and leaving AtomicBoolean is thread-safe against the test that was written for SOLR-7113; so AtomicBoolean gets the job done for both, even if it's not strictly necessary as far as application code goes.

**Comment by magibney** [comment]
Ok, I'm planning to merge this tomorrow. The most recent commits add some extra documentation, a test fix, clarity, robustness, and simplify a bit (IMO). I will keep an eye on the builds for any related failures.

File: solr/core/src/test/org/apache/solr/update/CustomTLogDirTest.java
Comment (magibney) [reply solr/core/src/test/org/apache/solr/update/CustomTLogDirTest.java]: thanks, removed this in 9dedb54121bb9822d8eeb2f301f47a2e09556cdd

**Comment by magibney** [comment]
Proposed CHANGES.txt:

```
* SOLR-16962: Restore ability to configure tlog directory. Includes substantial refactoring and clarification
  of logic and definitions around `ulog` and `tlog`. Refactoring largely focuses on UpdateLog and
  HdfsUpdateLog. There is no anticipated change in existing behavior, other than that the `dir` parameter
  to the `<updateLog>` element in `solrconfig.xml`, which has long been silently ignored, will now be
  respected (Michael Gibney, David Smiley)
```

**Comment by dsmiley** [comment]
You might leave the "refactoring" aspect to the commit message and just put in CHANGES.txt what users ought to know.

**Comment by magibney** [comment]
> You might leave the "refactoring" aspect to the commit message and just put in CHANGES.txt what users ought to know.

Revised:
```
* SOLR-16962: Restore ability to configure tlog directory. Changes largely focus on UpdateLog and
  HdfsUpdateLog. There is no anticipated change in existing behavior, other than that the `dir` parameter
  to the `<updateLog>` element in `solrconfig.xml`, which has long been silently ignored, will now be
  respected (Michael Gibney, David Smiley)
```

**Comment by iamsanjay** [comment]
git bisect points to this PR. Require bit more attention to see whether this PR causing it or not.

**org.apache.solr.search.TestCollapseQParserPlugin.testMultiSort (:solr:core)**
  ```
  Test history: https://ge.apache.org/scans/tests?search.rootProjectNames=solr-root&tests.container=org.apache.solr.search.TestCollapseQParserPlugin&tests.test=testMultiSort http://fucit.org/solr-jenkins-reports/history-trend-of-recent-failures.html#series/org.apache.solr.search.TestCollapseQParserPlugin.testMultiSort
    Test output: /Users/sanjaydutt/Documents/solr/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.search.TestCollapseQParserPlugin.txt
    Reproduce with: ./gradlew :solr:core:test --tests "org.apache.solr.search.TestCollapseQParserPlugin.testMultiSort" -Ptests.jvms=4 -Ptests.haltonfailure=false "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:+UseParallelGC -XX:ActiveProcessorCount=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=F8EF1414D2733583 -Ptests.multiplier=2 -Ptests.badapples=false -Ptests.file.encoding=US-ASCII
```

**Comment by jkrauss82** [comment]
Thanks for merging this.

I am trying to upgrade our Solr cloud node from Solr `9.5.0` to `9.8.0`, which has this PR merged. However, something does not seem to work as expected as Solr is not able to initialize the core, neither with tlogs copied to the new location, nor with an empty location.

The error encountered is given in the log dump below.

I had previously configured the ulog location like this:

```xml
<updateLog>
    <str name="dir">${solr.ulog.dir:/solr/logs/author_search}</str>
    <int name="numVersionBuckets">${solr.ulog.numVersionBuckets:65536}</int>
</updateLog>
```

When Solr is starting, it seems to recognize the configuration but aborts with a message about missing file permissions. The location `/solr/logs/author_search` exists, belongs to user `solr:solr` and has permissions `0750` (I have even tried with 0777).

**Question: what is the target folder structure Solr creates under `/solr/logs/author_search` and what could cause the error I am encountering?**

My suspicion is I am doing something wrong in regards to the folder structure.

I would like to carry over the old tlogs from 9.5.0 to have a seemless transition but I am stuck with this problem for now.

**I can provide more logs or help debugging upon request.**

### Solr version etc. from log

```txt
head -n 20 /var/log/solr/solr.log.1
2025-01-23 12:31:29.519 INFO  (main) [c: s: r: x: t:] o.e.j.s.Server jetty-10.0.22; built: 2024-06-27T16:03:51.502Z; git: 5c8471e852d377fd726ad9b1692c35ffc5febb09; jvm 21.0.4+8-LTS-274
2025-01-23 12:31:29.748 WARN  (main) [c: s: r: x: t:] o.e.j.u.DeprecationWarning Using @Deprecated Class org.eclipse.jetty.servlet.listener.ELContextCleaner
2025-01-23 12:31:29.770 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider Using logger factory org.apache.logging.slf4j.Log4jLoggerFactory
2025-01-23 12:31:29.775 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider  ___      _       Welcome to Apache Solr™ version 9.8.0
2025-01-23 12:31:29.775 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider / __| ___| |_ _   Starting in cloud mode on port 8983
2025-01-23 12:31:29.775 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider \__ \/ _ \ | '_|  Install dir: /opt/solr-9.8.0
2025-01-23 12:31:29.776 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider |___/\___/_|_|    Start time: 2025-01-23T11:31:29.776093489Z
2025-01-23 12:31:29.778 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider Solr started with "-XX:+CrashOnOutOfMemoryError" that will crash on any OutOfMemoryError exception. The cause of the OOME will be logged in the crash file at the following path: /var/log/solr/jvm_crash_3999500.log
2025-01-23 12:31:29.865 INFO  (main) [c: s: r: x: t:] o.a.l.s.MemorySegmentIndexInputProvider Using MemorySegmentIndexInput and native madvise support with Java 21 or later; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
2025-01-23 12:31:29.890 INFO  (main) [c: s: r: x: t:] o.a.l.i.v.PanamaVectorizationProvider Java vector incubator API enabled; uses preferredBitSize=256; FMA enabled
2025-01-23 12:31:29.892 INFO  (main) [c: s: r: x: t:] o.a.s.s.CoreContainerProvider Solr Home: /solr/data (source: system property: solr.solr.home)
2025-01-23 12:31:29.911 WARN  (main) [c: s: r: x: t:] o.a.s.c.c.SolrZkClient Using default ZkCredentialsInjector. ZkCredentialsInjector is not secure, it creates an empty list of credentials which leads to 'OPEN_ACL_UNSAFE' ACLs to Zookeeper nodes
2025-01-23 12:31:30.043 INFO  (main) [c: s: r: x: t:] o.a.s.c.c.ConnectionManager Waiting up to 30000ms for client to connect to ZooKeeper
2025-01-23 12:31:30.183 INFO  (zkCallback-1-thread-1) [c: s: r: x: t:] o.a.s.c.c.ConnectionManager zkClient has connected
2025-01-23 12:31:30.183 INFO  (main) [c: s: r: x: t:] o.a.s.c.c.ConnectionManager Client is connected to ZooKeeper
2025-01-23 12:31:30.183 WARN  (main) [c: s: r: x: t:] o.a.s.c.c.SolrZkClient Using default ZkACLProvider. DefaultZkACLProvider is not secure, it creates 'OPEN_ACL_UNSAFE' ACLs to Zookeeper nodes
2025-01-23 12:31:30.362 INFO  (main) [c: s: r: x: t:] o.a.s.c.NodeConfig Loading solr.xml from SolrHome (not found in ZooKeeper)
2025-01-23 12:31:30.362 INFO  (main) [c: s: r: x: t:] o.a.s.c.SolrXmlConfig solr.xml not found in SOLR_HOME, using built-in default
2025-01-23 12:31:30.362 INFO  (main) [c: s: r: x: t:] o.a.s.c.SolrXmlConfig Loading solr.xml from /opt/solr-9.8.0/server/solr/solr.xml
2025-01-23 12:31:30.389 INFO  (main) [c: s: r: x: t:] o.a.s.c.SolrResourceLoader Added 1 libs to classloader, from paths: [/opt/solr-9.8.0/lib]
```

### Core loading logs:

```txt
$ grep author_search /var/log/solr/solr.log.1

2025-01-23 12:31:32.108 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.c.SolrConfig Using Lucene MatchVersion: 9.9.0
2025-01-23 12:31:32.113 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.s.IndexSchemaSchema name=default-config
2025-01-23 12:31:32.120 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.s.IndexSchemaLoaded schema default-config/1.6 with uniqueid field id
2025-01-23 12:31:32.120 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.c.CoreContainer Creating SolrCore 'author_search_shard1_replica_n2' using configuration from configset author_search, trusted=false
2025-01-23 12:31:32.156 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.c.SolrCore Opening new SolrCore at [/solr/data/author_search_shard1_replica_n2], dataDir=[/solr/data/author_search_shard1_replica_n2/data/]
2025-01-23 12:31:32.669 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.j.SolrRequestAuthorizer Creating a new SolrRequestAuthorizer
2025-01-23 12:31:32.694 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.u.UpdateHandler Using UpdateLog implementation: org.apache.solr.update.UpdateLog
2025-01-23 12:31:32.694 WARN  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.u.UpdateLog numVersionBuckets is obsolete
2025-01-23 12:31:32.698 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.u.UpdateLog Initializing UpdateLog: dataDir=/solr/logs/author_search defaultSyncLevel=FLUSH numRecordsToKeep=100 maxNumLogsToKeep=10
2025-01-23 12:31:32.699 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.u.CommitTracker Hard AutoCommit: if uncommitted for 300000ms; if tlog file size has exceeded 134217728 bytes
2025-01-23 12:31:32.700 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.u.CommitTracker Soft AutoCommit: if uncommitted for 60000ms;
2025-01-23 12:31:32.701 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.c.SolrCore CLOSING SolrCore org.apache.solr.core.SolrCore@4e4e0e7e author_search_shard1_replica_n2
2025-01-23 12:31:32.702 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.m.SolrMetricManager Closing metric reporters for registry=solr.core.author_search.shard1.replica_n2 tag=SolrCore@4e4e0e7e
2025-01-23 12:31:32.726 INFO  (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c:author_search s:shard1 r:core_node4 x:author_search_shard1_replica_n2 t:] o.a.s.m.SolrMetricManager Closing metric reporters for registry=solr.collection.author_search.shard1.leader tag=SolrCore@4e4e0e7e
2025-01-23 12:31:32.731 ERROR (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c: s: r: x: t:] o.a.s.c.CoreContainer SolrCore failed to load on startup => org.apache.solr.common.SolrException: Unable to create core [author_search_shard1_replica_n2]
org.apache.solr.common.SolrException: Unable to create core [author_search_shard1_replica_n2]
Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "/solr/logs/author_search/author_search_shard1_replica_n2" "read")
```

### Java stacktrace

```txt
2025-01-23 12:31:32.731 ERROR (coreLoadExecutor-20-thread-1-processing-solr-fsn-2:8983_solr) [c: s: r: x: t:] o.a.s.c.CoreContainer SolrCore failed to load on startup => org.apache.solr.common.SolrException: Unable to create core [author_search_shard1_replica_n2]
at org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1778)
org.apache.solr.common.SolrException: Unable to create core [author_search_shard1_replica_n2]
at org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1778) ~[?:?]
at org.apache.solr.core.CoreContainer.lambda$loadInternal$12(CoreContainer.java:1080) ~[?:?]
at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:212) ~[metrics-core-4.2.26.jar:4.2.26]
at org.apache.solr.common.util.ExecutorUtil$MDCAwareThreadPoolExecutor.lambda$execute$0(ExecutorUtil.java:380) ~[?:?]
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: org.apache.solr.common.SolrException: Error Instantiating Update Handler, solr.DirectUpdateHandler2 failed to instantiate org.apache.solr.update.UpdateHandler
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1229) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1058) ~[?:?]
at org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1745) ~[?:?]
... 6 more
Caused by: org.apache.solr.common.SolrException: Error Instantiating Update Handler, solr.DirectUpdateHandler2 failed to instantiate org.apache.solr.update.UpdateHandler
at org.apache.solr.core.SolrCore.createInstance(SolrCore.java:969) ~[?:?]
at org.apache.solr.core.SolrCore.createUpdateHandler(SolrCore.java:1045) ~[?:?]
at org.apache.solr.core.SolrCore.initUpdateHandler(SolrCore.java:1289) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1175) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1058) ~[?:?]
at org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1745) ~[?:?]
... 6 more
Caused by: java.lang.reflect.InvocationTargetException
at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74) ~[?:?]
at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502) ~[?:?]
at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486) ~[?:?]
at org.apache.solr.core.SolrCore.createInstance(SolrCore.java:948) ~[?:?]
at org.apache.solr.core.SolrCore.createUpdateHandler(SolrCore.java:1045) ~[?:?]
at org.apache.solr.core.SolrCore.initUpdateHandler(SolrCore.java:1289) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1175) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1058) ~[?:?]
at org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1745) ~[?:?]
... 6 more
Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "/solr/logs/author_search/author_search_shard1_replica_n2" "read")
at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:488) ~[?:?]
at java.base/java.security.AccessController.checkPermission(AccessController.java:1071) ~[?:?]
at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411) ~[?:?]
at java.base/java.lang.SecurityManager.checkRead(SecurityManager.java:742) ~[?:?]
at java.base/sun.nio.fs.UnixPath.checkRead(UnixPath.java:789) ~[?:?]
at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributesIfExists(UnixFileSystemProvider.java:182) ~[?:?]
at java.base/java.nio.file.Files.isDirectory(Files.java:2319) ~[?:?]
at org.apache.lucene.store.FSDirectory.<init>(FSDirectory.java:127) ~[?:?]
at org.apache.lucene.store.MMapDirectory.<init>(MMapDirectory.java:244) ~[?:?]
at org.apache.lucene.store.MMapDirectory.<init>(MMapDirectory.java:173) ~[?:?]
at org.apache.lucene.store.FSDirectory.open(FSDirectory.java:162) ~[?:?]
at org.apache.solr.core.NRTCachingDirectoryFactory.create(NRTCachingDirectoryFactory.java:53) ~[?:?]
at org.apache.solr.core.CachingDirectoryFactory.get(CachingDirectoryFactory.java:410) ~[?:?]
at org.apache.solr.update.UpdateLog.initTlogDir(UpdateLog.java:570) ~[?:?]
at org.apache.solr.update.UpdateLog.init(UpdateLog.java:469) ~[?:?]
at org.apache.solr.update.UpdateHandler.initUlog(UpdateHandler.java:193) ~[?:?]
at org.apache.solr.update.DirectUpdateHandler2.<init>(DirectUpdateHandler2.java:160) ~[?:?]
at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62) ~[?:?]
at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502) ~[?:?]
at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486) ~[?:?]
at org.apache.solr.core.SolrCore.createInstance(SolrCore.java:948) ~[?:?]
at org.apache.solr.core.SolrCore.createUpdateHandler(SolrCore.java:1045) ~[?:?]
at org.apache.solr.core.SolrCore.initUpdateHandler(SolrCore.java:1289) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1175) ~[?:?]
at org.apache.solr.core.SolrCore.<init>(SolrCore.java:1058) ~[?:?]
at org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1745) ~[?:?]
... 6 more
```

**Pull request merged**
